### PR TITLE
JBPM-5536: Lienzo Panel & implementations (growable,resizable)

### PR DESCRIPTION
**Pull request by romartin**
Hey @hasys 

This provides some refactor on the Lienzo panel's and also two different implementations: growable & resizable.

See https://issues.jboss.org/browse/JBPM-5536

PS: Fixed the regression tests, but still working on tests for new stuff. Sorry will provide ASAP, but at least you can start the review.

Relates to:
https://github.com/kiegroup/kie-wb-common/pull/2236/
https://github.com/kiegroup/lienzo-tests/pull/53

Thanks!

### CODE REVIEW

File: src/main/java/com/ait/lienzo/client/core/shape/wires/WiresManager.java [hunk 1]
```diff
@@ -101,6 +101,19 @@ public static final WiresManager get(Layer layer)
         return manager;
     }
 
+    public static boolean isManaged(Layer layer)
+    {
+        final String uuid = layer.uuid();
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/core/shape/wires/WiresManager.java]: Variable 'uuid' is never used

File: src/main/java/com/ait/lienzo/client/core/shape/wires/WiresManager.java [hunk 2]
```diff
@@ -101,6 +101,19 @@ public static final WiresManager get(Layer layer)
...
+
+        WiresManager manager = MANAGER_MAP.get(layer.uuid());
+
+        if (null != manager)
+        {
+            return true;
+        }
+        return false;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/core/shape/wires/WiresManager.java]: I guess entire method can be like:

```java
return MANAGER_MAP.get(layer.uuid()) != null;
```

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelBoundsChangedEvent.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package com.ait.lienzo.client.widget.panel.event;
+
+import com.ait.lienzo.client.widget.panel.AbstractLienzoPanelEvent;
+
+public class LienzoPanelBoundsChangedEvent
+        extends AbstractLienzoPanelEvent<LienzoPanelBoundsChangedEventHandler>
+{
+    public static final Type<LienzoPanelBoundsChangedEventHandler> TYPE = new Type<LienzoPanelBoundsChangedEventHandler>();
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelBoundsChangedEvent.java]: Explicit type argument `LienzoPanelBoundsChangedEventHandler` can be replaced with <>

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelResizeEvent.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package com.ait.lienzo.client.widget.panel.event;
+
+import com.ait.lienzo.client.widget.panel.AbstractLienzoPanelEvent;
+
+public class LienzoPanelResizeEvent
+        extends AbstractLienzoPanelEvent<LienzoPanelResizeEventHandler>
+{
+    public static final Type<LienzoPanelResizeEventHandler> TYPE = new Type<LienzoPanelResizeEventHandler>();
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelResizeEvent.java]: Explicit type argument `LienzoPanelResizeEventHandler` can be replaced with <>

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelResizeEvent.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
+package com.ait.lienzo.client.widget.panel.event;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelResizeEvent.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelBoundsChangedEventHandler.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package com.ait.lienzo.client.widget.panel.event;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelBoundsChangedEventHandler.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelBoundsChangedEvent.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package com.ait.lienzo.client.widget.panel.event;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelBoundsChangedEvent.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/LienzoPanel.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package com.ait.lienzo.client.widget.panel;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/LienzoPanel.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/LienzoBoundsPanel.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
+package com.ait.lienzo.client.widget.panel;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/LienzoBoundsPanel.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/BoundsProvider.java [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package com.ait.lienzo.client.widget.panel;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/BoundsProvider.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/AbstractLienzoPanelEvent.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
+package com.ait.lienzo.client.widget.panel;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/AbstractLienzoPanelEvent.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/impl/ScalablePanel.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
+package com.ait.lienzo.client.widget.panel.impl;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/impl/ScalablePanel.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/impl/PreviewPanel.java [hunk 13]
```diff
@@ -0,0 +1,357 @@
+package com.ait.lienzo.client.widget.panel.impl;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/impl/PreviewPanel.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/impl/PreviewLayerDecorator.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
+package com.ait.lienzo.client.widget.panel.impl;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/impl/PreviewLayerDecorator.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/impl/PreviewLayer.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
+package com.ait.lienzo.client.widget.panel.impl;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/impl/PreviewLayer.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/impl/BoundsProviderFactory.java [hunk 16]
```diff
@@ -0,0 +1,215 @@
+package com.ait.lienzo.client.widget.panel.impl;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/impl/BoundsProviderFactory.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScrollEventHandler.java [hunk 17]
```diff
@@ -0,0 +1,8 @@
+package com.ait.lienzo.client.widget.panel.event;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScrollEventHandler.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScrollEvent.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
+package com.ait.lienzo.client.widget.panel.event;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScrollEvent.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScaleChangedEventHandler.java [hunk 19]
```diff
@@ -0,0 +1,8 @@
+package com.ait.lienzo.client.widget.panel.event;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScaleChangedEventHandler.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScaleChangedEvent.java [hunk 20]
```diff
@@ -0,0 +1,34 @@
+package com.ait.lienzo.client.widget.panel.event;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScaleChangedEvent.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelResizeEventHandler.java [hunk 21]
```diff
@@ -0,0 +1,8 @@
+package com.ait.lienzo.client.widget.panel.event;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelResizeEventHandler.java]: Missing license.

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScaleChangedEvent.java [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+
+import com.ait.lienzo.client.core.types.Point2D;
+import com.ait.lienzo.client.widget.panel.AbstractLienzoPanelEvent;
+
+public class LienzoPanelScaleChangedEvent
+        extends AbstractLienzoPanelEvent<LienzoPanelScaleChangedEventHandler>
+{
+    public static final Type<LienzoPanelScaleChangedEventHandler> TYPE = new Type<LienzoPanelScaleChangedEventHandler>();
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScaleChangedEvent.java]: Explicit type argument `LienzoPanelScaleChangedEventHandler` can be replaced with <>

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScrollEvent.java [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+
+import com.ait.lienzo.client.widget.panel.AbstractLienzoPanelEvent;
+
+public class LienzoPanelScrollEvent
+        extends AbstractLienzoPanelEvent<LienzoPanelScrollEventHandler>
+{
+    public static final Type<LienzoPanelScrollEventHandler> TYPE = new Type<LienzoPanelScrollEventHandler>();
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScrollEvent.java]: Explicit type argument `LienzoPanelScrollEventHandler` can be replaced with <>

File: src/main/java/com/ait/lienzo/client/widget/panel/impl/LienzoPanelHandlerManager.java [hunk 24]
```diff
@@ -0,0 +1,889 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * This class is the center for all canvas events.
+ * <p/>
+ * It can not and should not be used directly as it is used internally by the toolkit to handle and dispatch {@link
+ * com.ait.lienzo.client.core.shape.Shape} events.
+ */
+public final class LienzoPanelHandlerManager
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/impl/LienzoPanelHandlerManager.java]: New class doesn't covered by tests.

File: src/main/java/com/ait/lienzo/client/widget/panel/impl/LienzoPanelImpl.java [hunk 25]
```diff
@@ -0,0 +1,551 @@
...
+
+/**
+ * LienzoPanel acts as a Container for a {@link com.ait.lienzo.client.core.shape.Viewport}.
+ * <p/>
+ * <ul> <li>An application will typically be composed of one or more LienzoPanels.</li> <li>A LienzoPanel takes width
+ * and height as input parameters.</li> <li>A {@link com.ait.lienzo.client.core.shape.Viewport} will contain one main
+ * {@link com.ait.lienzo.client.core.shape.Scene}</li> <li>The main {@link com.ait.lienzo.client.core.shape.Scene} can
+ * contain multiple {@link com.ait.lienzo.client.core.shape.Layer}.</li> </ul>
+ */
+public class LienzoPanelImpl extends LienzoPanel<LienzoPanelImpl>
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/impl/LienzoPanelImpl.java]: New class file doesn't covered by tests.

File: src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollablePanel.java [hunk 26]
```diff
@@ -0,0 +1,451 @@
...
+    @Override
+    public void onResize()
+    {
+        scheduleDeferred(new Scheduler.ScheduledCommand()
+        {
+            @Override
+            public void execute()
+            {
+                onSroll();
+                ;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollablePanel.java]: Useless semicolon.

File: src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollablePanel.java [hunk 27]
```diff
@@ -0,0 +1,451 @@
...
+    {
+        final Transform transform = getLayer().getViewport().getTransform();
+        m_events.fireEvent(new LienzoPanelScaleChangedEvent(new Point2D(transform.getScaleX(),
+                                                                        transform.getScaleY())));
+    }
+
+    public void updateSize()
+    {
+        final Element parentElement = getElement().getParentElement();
+        final Integer width         = parentElement.getOffsetWidth();
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollablePanel.java]: Type may be primitive

File: src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollablePanel.java [hunk 28]
```diff
@@ -0,0 +1,451 @@
...
+        final Integer height        = parentElement.getOffsetHeight();
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollablePanel.java]: Type may be primitive

File: src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java [hunk 29]
```diff
@@ -0,0 +1,96 @@
...
+
+    private final ScrollablePanelHandler scrollHandler;
+
+    ScrollPosition(final ScrollablePanelHandler scrollHandler) {
+        this.scrollHandler = scrollHandler;
+    }
+
+    double currentRelativeX() {
+
+        final Double delta = deltaX();
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java]: Type may be primitive

File: src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java [hunk 30]
```diff
@@ -0,0 +1,96 @@
...
+
+        return delta == 0d ? 0d : 100 * currentX() / delta;
+    }
+
+    double currentRelativeY() {
+
+        final Double delta = deltaY();
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java]: Type may be primitive

File: src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java [hunk 31]
```diff
@@ -0,0 +1,96 @@
...
+
+        return delta == 0d ? 0d : 100 * currentY() / delta;
+    }
+
+    double currentPositionX(final Double level) {
+
+        final Double position = deltaX() * level / 100;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java]: Type may be primitive

File: src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java [hunk 32]
```diff
@@ -0,0 +1,96 @@
...
+
+        return -(bounds().minBoundX() + position);
+    }
+
+    double currentPositionY(final Double level) {
+
+        final Double position = deltaY() * level / 100;
```
Comment (hasys) [inline src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java]: Type may be primitive

File: src/main/java/com/ait/lienzo/client/core/shape/wires/WiresManager.java (hunk 1 above)
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/core/shape/wires/WiresManager.java]: done

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelBoundsChangedEvent.java (hunk 3 above)
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelBoundsChangedEvent.java]: done

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScaleChangedEvent.java (hunk 22 above)
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScaleChangedEvent.java]: done

File: src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScrollEvent.java (hunk 23 above)
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/widget/panel/event/LienzoPanelScrollEvent.java]: - LienzoPanelImpl

File: src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollablePanel.java (hunk 27 above)
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollablePanel.java]: done

File: src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollablePanel.java (hunk 28 above)
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollablePanel.java]: done

File: src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java (hunk 29 above)
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java]: done

File: src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java (hunk 31 above)
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java]: done

File: src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java (hunk 32 above)
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/widget/panel/scrollbars/ScrollPosition.java]: done

**Comment by romartin** [comment]
squashed commits manually before merging

**Pull request merged**